### PR TITLE
Browserify

### LINKED ISSUE #24: stream.js and lib/index.js are not in sync
[stream.js](https://github.com/dionyziz/stream.js/blob/master/stream.js) does not have `zip`, `drop` etc. that [index.js](https://github.com/dionyziz/stream.js/blob/master/lib/index.js) has.

This can be resolved manually, automatically or by uses [UMD](https://github.com/umdjs/umd) in index.js and delete stream.js.

### PR DESCRIPTION
**Pull request by dotnetCarpenter**
Since over the wire minifying is deemed unsafe...
Switch to browserify which gives us AMD support (we already have CommonJS) and a minifier that also produce a source map. Fixes #24.

I think this should be a minor release as AMD support is a new feature. I need to change the website to use npm as build tool instead of make. The idea is to somehow use this minifier in the website repo and compress the stream.min.js and stream.min.js.map in a zip file for download via the website.

I also added a simple introduction in the README.

### CODE REVIEW

**Comment by dotnetCarpenter** [comment]
hehe updating jasmine brings up our code coverage

**Comment by dotnetCarpenter** [comment]
I just found [rollup](http://rollupjs.org/) tonight. I bundles es6 modules to different formats. I've been looking for that for some time. Actually I could talk heaps about it's many great features but it's late and I'm going to bed. Suffice to say, it's a step up from browserify and I would like to make the switch when I have time. On the plus side, [PR #2 in website](https://github.com/dionyziz/stream.js-website/pull/2) does not need to change, if we switch.

File: stream.js [hunk 1]
```diff
@@ -1,3 +1,4 @@
+(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.Stream = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
```
Comment (dionyziz) [inline stream.js]: Why do we need these here?

File: stream.js [hunk 2]
```diff
@@ -272,3 +306,22 @@ Stream.equals = function ( stream1, stream2 ) {
...
+};
+Stream.cycle = function( array ) {
+  var promise_generator = function( array, index) {
+    if( index >= array.length ) index = 0;
+    return function() {
+      return new Stream( array[index], promise_generator(array, index + 1 ) );
+    };
+  };
+  return new Stream( array[0], promise_generator( array, 1 ) );
+};
```
Comment (dionyziz) [inline stream.js]: The changes to add iterate, cycle, repeat, continually, dropWhile look quite nice, but I think they can be in a different PR so that we can discuss these two proposal changes independently.

However, since you did all this work, we can keep them here if you don't feel like splitting up – we'll probably merge both anyway.

File: stream.js [hunk 3]
```diff
@@ -272,3 +306,22 @@ Stream.equals = function ( stream1, stream2 ) {
...
+
+module.exports = Stream;
+
+},{}]},{},[1])(1)
+});
```
Comment (dionyziz) [inline stream.js]: New line at end of file

File: stream.js [hunk 4]
```diff
@@ -272,3 +306,22 @@ Stream.equals = function ( stream1, stream2 ) {
         return Stream.equals( stream1.tail(), stream2.tail() );
     }
 };
+Stream.iterate = function ( x, f ) {
+  return new Stream( x, function () {
+    return Stream.iterate( f( x ), f);
+  } );
+};
+Stream.cycle = function( array ) {
+  var promise_generator = function( array, index) {
```
Comment (dionyziz) [inline stream.js]: coding style

File: stream.js [hunk 5]
```diff
@@ -192,7 +213,14 @@ Stream.prototype = {
         // create clone/a contructor which accepts a stream?
         return new Stream( self.headValue, self.tailPromise );
     },
-    member: function( x ){
+    dropWhile: function( condition ) {
+      if(this.empty()) return new Stream();
+      var self = this;
+
+      while( condition( self.head() ) ) self = self.tail();
```
Comment (dionyziz) [inline stream.js]: let's maintain `{}` around one-line bodies for consistency

File: stream.js [hunk 6]
```diff
@@ -192,7 +213,14 @@ Stream.prototype = {
         // create clone/a contructor which accepts a stream?
         return new Stream( self.headValue, self.tailPromise );
     },
-    member: function( x ){
+    dropWhile: function( condition ) {
+      if(this.empty()) return new Stream();
```
Comment (dionyziz) [inline stream.js]: conding style

File: stream.js [hunk 7]
```diff
@@ -224,14 +252,20 @@ Stream.prototype = {
...
+    return Stream.repeat( 1 );
 };
 Stream.makeNaturalNumbers = function() {
     return new Stream( 1, function () {
         return Stream.makeNaturalNumbers().add( Stream.makeOnes() );
     } );
 };
+Stream.repeat = function( element ) {
+    return Stream.continually( function() { return element; } );
+};
```
Comment (dionyziz) [inline stream.js]: Nice, very useful.

File: stream.js [hunk 8]
```diff
@@ -272,3 +306,22 @@ Stream.equals = function ( stream1, stream2 ) {
         return Stream.equals( stream1.tail(), stream2.tail() );
     }
 };
+Stream.iterate = function ( x, f ) {
+  return new Stream( x, function () {
+    return Stream.iterate( f( x ), f);
+  } );
+};
+Stream.cycle = function( array ) {
```
Comment (dionyziz) [inline stream.js]: I don't see why this can't also accept streams.

File: stream.js [hunk 9]
```diff
@@ -272,3 +306,22 @@ Stream.equals = function ( stream1, stream2 ) {
         return Stream.equals( stream1.tail(), stream2.tail() );
     }
 };
+Stream.iterate = function ( x, f ) {
+  return new Stream( x, function () {
+    return Stream.iterate( f( x ), f);
```
Comment (dionyziz) [inline stream.js]: style

File: stream.js [hunk 10]
```diff
@@ -272,3 +306,22 @@ Stream.equals = function ( stream1, stream2 ) {
         return Stream.equals( stream1.tail(), stream2.tail() );
     }
 };
+Stream.iterate = function ( x, f ) {
```
Comment (dionyziz) [inline stream.js]: Nice one, looks super useful.

File: stream.js (hunk 1 above)
Comment (dotnetCarpenter) [inline stream.js]: That is the UMD stuff. You can use require via CommonJS, define via AMD og just plain old global variable to consume stream.js. I don't know why browserify makes it look so bloated. Rollup insert this for comparison:
```js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
    typeof define === 'function' && define.amd ? define(factory) :
    global.Stream = factory();
}(this, function () { 'use strict';
  // stream.js
  return Stream;
}));
```

I'm still looking into [rollup](https://github.com/rollup/rollup). There is a couple of options to how you want users to be able to consume stream.js. I'll make an issue when I've done a bit more digging.

**Comment by dotnetCarpenter** [comment]
I'm very happy you like the additional methods but I can't take credit. These changes was there all along, made by @tech-no-crat, last year. We just had two different versions of stream.js, hence issue #24. Browserify takes the node (CommonJS) compliant version and creates a UMD version for everyone to use.

File: stream.js (hunk 3 above)
Comment (dotnetCarpenter) [inline stream.js]: I'm afraid that this file is generated by browserify and new line at end of file is not in my control.
